### DISCUSSION TITLE
What's the rationale behind a single feature setting multiple bits in RDKit fingerprints?

### DISCUSSION THREAD

**Discussion by AlanKerstjens** [rdkit/rdkit#4352] [Q&A] [answered]
Dear RDKit community,

I was looking into the RDKit fingerprints, and noticed that the free function to generate folded fingerprints (`RDKFingerprintMol()`), as well as the corresponding generator (`getRDKitFPGenerator()`) take as an argument the `nBitsPerHash`/`nBitsPerFeature`. `getUnfoldedRDKFingerprintMol()` lacks this argument, but for unfolded fingerprints the same behavior can still be achieved through the `RDKitFPGenerator`. 

I saw in the RDKit book that if this number is greater than 1, "each subgraph will set more than one bit", with the additional bits being generated "by seeding a random number generator with the original raw bit ID and generating the appropriate number of random numbers".

What's the advantage of adding these additional "random" bits to the fingerprint? In terms of usage guidelines, are there any differences between folded and unfolded fingerprints?

**Comment by greglandrum** [comment, marked as answer]
Hi @AlanKerstjens,
The option to generate multiple bits for the RDKit fingerprint is there primarilty for historical reasons - we modeled the fingerprint on the Daylight fingerprint and that set multiple bits per feature. 

One reason to set multiple bits per feature is that doing so can decrease the impact of bit collisions caused by different features which hash to different "unfolded" values being folded to the same bit id. Here's a concrete example of that.
Start with a molecule which has two features which would fold to the same bit if we fold a fingerprint to 1024 bits.
Here's the molecule with the bonds labeled:
![image](https://user-images.githubusercontent.com/540511/127091564-f480d1b2-01db-4f23-b7d1-8a4207d96c38.png)

Now let's look for unfolded features which fold to the same bit ID:
```
>>> m = Chem.MolFromSmiles('O=C1NCC1C')
>>> bi2 = {}
>>> fp = Chem.UnfoldedRDKFingerprintCountBased(m,bitInfo=bi2)
>>> seen = {}
>>> for k in bi2:
...     fk = k%1024
...     if fk in seen:
...         print(fk,k,bi2[k],seen[fk][0],seen[fk][1])
...     else:
...         seen[fk] = (k,bi2[k])
... 
336 3170856272 [[1, 2, 3, 5, 4]] 1480085840 [[0, 5, 4], [0, 5, 3]]
```
so in this fingerprint the features defined by bond sets [1,2,3,5,4], [0,5,4], and [0,5,3] all end up setting bit 336 if the fingerprint is folded to be 1024 bits long. Note that the features defined by bond sets [0,5,4] and [0,5,3] are identical to each other as far as the RDKit fingerprint is concerned - both hash to unfolded bit ID 1480085840 - but that the feature defined by bond set [1,2,3,5,4] sets a different bit in the unfolded fingerprint - 3170856272.

Setting two bits per feature in the 1024 bit fingerprint shows that the two feature types collide at bit 336 (which we already saw), but that the second bit they each set do not collide:
```
>>> bi = {}
>>> fp = Chem.RDKFingerprint(m,fpSize=1024,bitInfo=bi)
>>> for k in bi:
...     if [0,5,4] in bi[k] or [1,2,3,5,4] in bi[k]:
...         print(k,bi[k])
... 
200 [[0, 5, 4], [0, 5, 3]]
336 [[0, 5, 4], [0, 5, 3], [1, 2, 3, 5, 4]]
998 [[1, 2, 3, 5, 4]]
```
So we end up with 3 bits set instead of 4. That's some loss of information, but it's not as bad as if we only set one bit per feature, which in this case would have completely lost the distinction between those two features by only setting bit 336.


This approach of setting multpiple bits per feature/item is also used in [locality sensitive hashing](https://en.wikipedia.org/wiki/Locality-sensitive_hashing) and [Bloom filters](https://en.wikipedia.org/wiki/Bloom_filter).

**Reply by AlanKerstjens** [reply]
Thanks for the comprehensive explanation @greglandrum! The problem of distinct features setting the same bit due to hashing collisions is of course not unique to RDKit fingerprints. For RDKit fingerprints the default is `nBitsPerFeature=2`, yet for Morgan fingerprints, as far as I can tell, we don't have this functionality at all. Is this also due to historical reasons to try to stay true to the ECFP fingerprints? What's your personal take on the usefulness of this feature?

**Reply by Dingyun-Huang** [reply]
> 200 [[0, 5, 4], [0, 5, 3]]
> 336 [[0, 5, 4], [0, 5, 3], [1, 2, 3, 5, 4]]
> 998 [[1, 2, 3, 5, 4]]

Hi there, just curious about does this mean 2 bits in the feature vector may represent the same fingerprint?
